aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -0,0 +1,366 @@
+
+#include <ATen/FunctionalTensorWrapper.h>
+
+#include <ATen/FunctionalInverses.h>
+#include <ATen/WrapDimUtils.h>
+#include <ATen/core/LegacyTypeDispatch.h>
+#include <c10/util/Exception.h>
+
+#include <c10/util/irange.h>
+
+namespace at {
+
+FunctionalTensorWrapper::FunctionalTensorWrapper(Tensor value)
+  : c10::TensorImpl(
+      c10::Storage(c10::make_intrusive<functionalization::FunctionalStorageImpl>(value.device(), value.numel(), value.dtype())),
+      c10::DispatchKeySet(DispatchKey::Functionalize),
+      value.dtype()
+    ),
+    value_(value),
+    // Note: "level" is a concept that we don't know how to compute in core.
+    // For now I'm retroactively setting this in functorch,
+    // but once Open Multiple Dispatch lands we should be able to calculate this in core.
+    level_(-1)
+{
+  // shallow_copy_from overwrites the storage and dispatch keyset...
+  auto functional_storage = storage_;
+  shallow_copy_from(value.getIntrusivePtr());
+  storage_ = functional_storage;
+  // TODO: should FunctionalTensorWrapper set_storage_access_should_throw()?
+  // Their storage is similar to meta tensors, which does not throw.
+  // However, users might expect to legally use storage() inside of a functionalized program, and this could be a foot gun.
+  storage_access_should_throw_ = false;
+  TORCH_INTERNAL_ASSERT(value_.defined());
+  // TODO: test using the python key to trace a function that uses the functionalization pass.
+  // I think I need this to make sure that replace_() calls hit the python key.
+  key_set_ = (value_.key_set() & c10::DispatchKeySet(c10::DispatchKeySet::FULL_AFTER, c10::DispatchKey::FuncTorchDynamicLayerBackMode))
```

Comment:
Yeah, I'm not sure about this part. If you want to keep the Python key around (and I'm not sure you do), just say that, instead of grabbing everything as you do here. A CPU dispatch on a FunctionalTensorWrapper is definitely not going to work

Reply:
sounds good. I think I'm leaning towards not worry about the python key use case for now in this PR.